reinforcement_learning/rl_game_server_autopilot/ap-server/app.py

```diff
@@ -0,0 +1,228 @@
+cloudwatch_region = 'us-west-2'
+cloudwatch_cli = boto3.client('cloudwatch',region_name=cloudwatch_region)
+
+app = Chalice(app_name='autopilot-server')
+
+#TBD Export to environment variables
+endpoint_name ="sagemaker-tensorflow-serving-2019-09-23-20-53-20-237"
+namespace = "autopilot"
+gs_inventory_url = "https://3vwufqstq5.execute-api.us-west-2.amazonaws.com/api/currsine24h"
+
+
+supported_calls={"SupportedCalls":
+                  {
+                     "/predict":"get game-server predictions",
+                     "/currtime":"get the current time in the cloud",
+                     "/currsine1h":"get sine value of 1-hour full positive sine cycle",
+                     "/currsine24h":"get sine value of 24-hour full positive sine cycle",
+                  }
+                }
+
+@app.route('/')
+def index():
+    return supported_calls
+
+def deserializer(item,observations):
+    print('in deserializer '+str(item))
+    item=item.replace("[","")
+    item=item.replace("]","")
+    print('item='+str(item))
+    for num in item.split(" "):
+       if (num):
+         print('num='+str(num))
+         observations.append(float(num))
+
+def get_last_obs_arr(table):
+    print('in get_last_obs_arr {}'.format(table))
+    observations=[]
+    try:
+      response = table.get_item(
+      Key={
+        'key': 'observation'
+      }
+      )
+      x=json.dumps(response['Item'])
+      item = response['Item']['value']
+      deserializer(item,observations)
+      print('observations='+str(observations))
+      return observations
+    except Exception as e:
+      print('Error {}'.format(e))
+
+def put_latest_gs_inference(value,table):
+    print ('in put_latest_gs_inference='+str(value))
+    observations=get_last_obs_arr(table)
+    print('observations before {}'.format(observations))
+    try:
+      observations=observations[1:]
+      observations=np.append(observations,float(value))
+      observations=str(observations)
+      print ('observations after appending {}'.format(observations))
+    except Exception as e:
+      print('Error {}'.format(e))
+    table.put_item(
+       Item={
+          'key': 'observation',
+          'value': observations
+       }
+    )
+    return observations
+
+def populate_cloudwatch_metric(namespace,metric_value,metric_name):
+    print("in populate_cloudwatch_metric metric_value="+str(metric_value)+" metric_name="+metric_name)
+    response = cloudwatch_cli.put_metric_data(
+        Namespace=namespace,
+        MetricData=[
+           {
+              'MetricName': metric_name,
+              'Unit': 'None',
+              'Value': metric_value,
+           },
+        ]
+        )
+    print('response from cloud watch'+str(response))
+
+@app.route('/predict/{region}')
+def get_prediction(region):
+    print ('in predict for region {}'.format(region))
+    dynamodb = boto3.resource('dynamodb',region_name=region)
+    latest_observations_table = dynamodb.Table('latest_observations')
```

Comment:
Ideally we should create boto3 DynamoDb resource per instantiation of the app/webserver. 

What happens here is everytime get_prediction endpoint is called, we create a new connection to DDb, which is very resource exhaustive